### PR TITLE
add playbook for gcache cleaning

### PR DESCRIPTION
**Pull request by Xiaohua-Shen**

### CODE REVIEW

File: playbooks/fix-mysql-gcache.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+    - name: clean existing gcache.page files on secondary node
+      shell: rm -rf /var/lib/mysql/gcache.page.*
+      when: ucarp_result.rc == 1
+
+    - name: start mysql service on secondary node
+      service: name=mysql state=started 
+      when: ucarp_result.rc == 1
+
+    - name: wait 10 minutes
```
Comment (sandfish8) [inline playbooks/fix-mysql-gcache.yml]: why do we wait 10 min ?

File: playbooks/fix-mysql-gcache.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
+---
+- hosts: controller
```
Comment (sandfish8) [inline playbooks/fix-mysql-gcache.yml]: please add `  any_errors_fatal: true`

**Comment by sandfish8** [comment]
Can we please verify wsrep_cluster_size is 3 before we start and also at the end.

**Comment by sandfish8** [comment]
@Xiaohua-Shen:

@nirajdp and I spoke and he pointed out that the initial implementation would have run into issues running in parallel against both controllers at the same time.
While we could have added `serial: 1` that also didn't completely solve the problem.

I think an approach like below with multiple plays would work.  
( Please pardon me for just writing this out but it seemed more clear than doing psuedo code.  I have not tested this code.

```
---
# check wsrep_cluster_size
hosts: controller
any_errors_fatal: true
tasks:
  - name: check galera cluster size
    shell: mysql -e "SHOW STATUS LIKE'%wsrep_cluster_size%'" | grep 3

# clean the non ucarp host
hosts: controller
any_errors_fatal: true
tasks:
  - name: check for ucarp interface
    shell: ip a | grep ucarp > /dev/null 2>&1
    register: ucarp_result
    ignore_errors: true

  - name: stop mysql service
    service: name=mysql state=stopped
    when: ucarp_result.rc == 1

  - name: clean existing gcache.page files
    shell: rm -rf /var/lib/mysql/gcache.page.*
    when: ucarp_result.rc == 1

  - name: start mysql service
    service: name=mysql state=started
    when: ucarp_result.rc == 1

# check wsrep_cluster_size
hosts: controller
any_errors_fatal: true
tasks:
  - name: check galera cluster size
    shell: mysql -e "SHOW STATUS LIKE'%wsrep_cluster_size%'" | grep 3

# move ucarp HA IP
hosts: controller
any_errors_fatal: true
tasks:
  - name: check for ucarp interface
    shell: ip a | grep ucarp > /dev/null 2>&1
    register: ucarp_result
    ignore_errors: true

  - name: find ucarp PID
    shell: pidof ucarp
    register: ucarp_pid
    when: ucarp_result.rc == 0

  - name: trigger ucarp failover
    shell: kill -USR2 {{ ucarp_pid.stdout }}
    when: ucarp_result.rc == 0

# clean the non ucarp host
hosts: controller
any_errors_fatal: true
tasks:
  - name: check for ucarp interface
    shell: ip a | grep ucarp > /dev/null 2>&1
    register: ucarp_result
    ignore_errors: true

  - name: stop mysql service
    service: name=mysql state=stopped
    when: ucarp_result.rc == 1

  - name: clean existing gcache.page files
    shell: rm -rf /var/lib/mysql/gcache.page.*
    when: ucarp_result.rc == 1

  - name: start mysql service
    service: name=mysql state=started
    when: ucarp_result.rc == 1

# check wsrep_cluster_size
hosts: controller
any_errors_fatal: true
tasks:
  - name: check galera cluster size
    shell: mysql -e "SHOW STATUS LIKE'%wsrep_cluster_size%'" | grep 3
```

**Comment by sandfish8** [comment]
@Xiaohua-Shen I broke into multiple plays b/c I though ansible would run in parallel for each play. However double checking http://docs.ansible.com/ansible/playbooks_strategies.html it looks like it's per task.
So I think your implementation would work (with some the cluster size checking thrown in).
Feel fee to just take that parts of you mine you like.

Thanks,
Sam

File: playbooks/fix-mysql-gcache.yml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  tasks:
+    - name: determine if it's a secondary node
```
Comment (sandfish8) [inline playbooks/fix-mysql-gcache.yml]:
```
 - name: check galera cluster size
    shell: mysql -e "SHOW STATUS LIKE'%wsrep_cluster_size%'" | grep 3
```

File: playbooks/fix-mysql-gcache.yml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+      when: ucarp_result.rc == 1
+
+    - name: clean existing gcache.page files on secondary node
+      shell: rm -rf /var/lib/mysql/gcache.page.*
+      when: ucarp_result.rc == 1
+
+    - name: start mysql service on secondary node
+      service: name=mysql state=started 
+      when: ucarp_result.rc == 1
+
```
Comment (sandfish8) [inline playbooks/fix-mysql-gcache.yml]:
```
 - name: check galera cluster size
    shell: mysql -e "SHOW STATUS LIKE'%wsrep_cluster_size%'" | grep 3
```

File: playbooks/fix-mysql-gcache.yml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+      service: name=mysql state=stopped 
+      when: ucarp_result.rc == 0
+
+    - name: clean existing gcache.page files on primary node
+      shell: rm -rf /var/lib/mysql/gcache.page.*
+      when: ucarp_result.rc == 0
+
+    - name: start mysql service on primary node
+      service: name=mysql state=started 
+      when: ucarp_result.rc == 0
```
Comment (sandfish8) [inline playbooks/fix-mysql-gcache.yml]:
```
 - name: check galera cluster size
    shell: mysql -e "SHOW STATUS LIKE'%wsrep_cluster_size%'" | grep 3
```

File: playbooks/fix-mysql-gcache.yml (hunk 1 above)
Comment (Xiaohua-Shen) [inline playbooks/fix-mysql-gcache.yml]: it's based on RTC 187425 related comment 2, seems operation team use this manual steps to clean the gcache pages. I guess the reason for the waiting is let 2nd node has enough time to sync the data from primary node.

**Comment by Xiaohua-Shen** [comment]
@sandfish8,  add cluster size check make the playbook more safe to run.  I update the code as you suggested, please review.

**Comment by nirajdp76** [comment]
retest

**Comment by nirajdp76** [comment]
retest

**Comment by nirajdp76** [comment]
retest

**Comment by nirajdp76** [comment]
retest

**Comment by nirajdp76** [comment]
retest

**Pull request merged**